src/xrpld/app/misc/detail/LendingHelpers.cpp

```diff
@@ -0,0 +1,1654 @@
+    Number const& principalOutstanding,
+    Number const& managementFeeOutstanding,
+    Number const& periodicPayment,
+    Number const& periodicRate,
+    std::uint32_t paymentRemaining,
+    TenthBips16 managementFeeRate)
+{
+    /*
+     * This function is derived from the XLS-66 spec, section 3.2.4.1.1 (Regular
+     * Payment)
+     */
+    XRPL_ASSERT_PARTS(
+        isRounded(asset, totalValueOutstanding, scale) &&
+            isRounded(asset, principalOutstanding, scale) &&
+            isRounded(asset, managementFeeOutstanding, scale),
+        "ripple::detail::computePaymentComponents",
+        "Outstanding values are rounded");
+    XRPL_ASSERT_PARTS(
+        paymentRemaining > 0,
+        "ripple::detail::computePaymentComponents",
+        "some payments remaining");
+    auto const roundedPeriodicPayment =
+        roundPeriodicPayment(asset, periodicPayment, scale);
+
+    LoanState const trueTarget = calculateRawLoanState(
+        periodicPayment, periodicRate, paymentRemaining - 1, managementFeeRate);
+    LoanState const roundedTarget = LoanState{
+        .valueOutstanding =
+            roundToAsset(asset, trueTarget.valueOutstanding, scale),
+        .principalOutstanding =
+            roundToAsset(asset, trueTarget.principalOutstanding, scale),
+        .interestDue = roundToAsset(asset, trueTarget.interestDue, scale),
+        .managementFeeDue =
+            roundToAsset(asset, trueTarget.managementFeeDue, scale)};
+    LoanState const currentLedgerState = calculateRoundedLoanState(
+        totalValueOutstanding, principalOutstanding, managementFeeOutstanding);
+
+    LoanDeltas deltas = currentLedgerState - roundedTarget;
+    deltas.nonNegative();
+
+    // Adjust the deltas if necessary for data integrity
+    XRPL_ASSERT_PARTS(
+        deltas.principalDelta <= currentLedgerState.principalOutstanding,
+        "ripple::detail::computePaymentComponents",
+        "principal delta not greater than outstanding");
+
+    deltas.principalDelta = std::min(
+        deltas.principalDelta, currentLedgerState.principalOutstanding);
+
+    XRPL_ASSERT_PARTS(
+        deltas.interestDueDelta <= currentLedgerState.interestDue,
+        "ripple::detail::computePaymentComponents",
+        "interest due delta not greater than outstanding");
+
+    deltas.interestDueDelta = std::min(
+        {deltas.interestDueDelta,
+         std::max(numZero, roundedPeriodicPayment - deltas.principalDelta),
+         currentLedgerState.interestDue});
+
+    XRPL_ASSERT_PARTS(
+        deltas.managementFeeDueDelta <= currentLedgerState.managementFeeDue,
+        "ripple::detail::computePaymentComponents",
+        "management fee due delta not greater than outstanding");
+
+    deltas.managementFeeDueDelta = std::min(
+        {deltas.managementFeeDueDelta,
+         roundedPeriodicPayment -
+             (deltas.principalDelta + deltas.interestDueDelta),
+         currentLedgerState.managementFeeDue});
```

Comment:
Perhaps, in the spirit of defensive programming it's worth adding a call to std::max, just to be extra safe that we won't end up with a negative fee payment
```suggestion
    deltas.managementFeeDueDelta = std::min(
        {deltas.managementFeeDueDelta,
         std::max(
             numZero,
             roundedPeriodicPayment -
                 (deltas.principalDelta + deltas.interestDueDelta)),
         currentLedgerState.managementFeeDue});
```